### PR TITLE
dcp-845 metadata spreadsheet in staging area

### PR DESCRIPTION
**Pull request by amnonkhen**
dcp-845
ebi-ait/dcp-ingest-central#845

### CODE REVIEW

File: broker/submissions/export_to_spreadsheet_service.py [hunk 1]
```diff
@@ -46,11 +46,28 @@ def export_and_save(self, submission_uuid: str, storage_dir: str):
         os.makedirs(spreadsheet_details.directory, exist_ok=True)
         workbook.save(spreadsheet_details.filepath)
 
+        # dcp-845 - spreadsheet in staging area
+        # create a file payload that adhere to the schema
+        # https://schema.humancellatlas.org/type/file/2.5.0/supplementary_file
+        spreadsheet_payload = self.build_supplementary_file_payload(spreadsheet_details)
+        # POST /submission/<id>/files
+        self.ingest_api.put(f'/submission/{submission_uuid}/files/', spreadsheet_payload)
```
Comment (amnonkhen) [inline broker/submissions/export_to_spreadsheet_service.py:54]: It needs to be POST

File: broker/submissions/export_to_spreadsheet_service.py [hunk 2]
```diff
@@ -46,11 +46,28 @@ def export_and_save(self, submission_uuid: str, storage_dir: str):
...
         self.logger.info(f'Done exporting spreadsheet for submission {submission_uuid}!')
 
+    def build_supplementary_file_payload(self, spreadsheet_details):
+        return {
+            "describedBy": "https://schema.humancellatlas.org/type/file/2.5.0/supplementary_file",
+            "schema_type": "file",
+            "file_core": {
+                "file_name": spreadsheet_details.filename,
+                "format": "xlsx"
+            }
```
Comment (amnonkhen) [inline broker/submissions/export_to_spreadsheet_service.py]: According to the [dcp spec](https://github.com/HumanCellAtlas/dcp2/blob/e4d843f75ec66a2d5ef1dadad0c259703b1bb085/docs/dcp2_system_design.rst#ingest-spreadsheets) the the spreadsheet should be have the metadata that describes its source.
For that the payload should include the `file_source` and `content_description` attributes. The latter can have an ontology reference, which will be [`data:2193`](https://ontology.archive.data.humancellatlas.org/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_2193) as [decided previously](https://app.zenhub.com/workspaces/dcp-ingest-product-development-5f71ca62a3cb47326bdc1b5c/issues/ebi-ait/dcp-ingest-central/845#issuecomment-1249230627).

```suggestion
            "file_core": {
                "file_name": spreadsheet_details.filename,
                "format": "xlsx"
                "file_source": "DCP/2 Ingest",
                "content_description": {
                    "text":"metadata spreadsheet",
                    "ontology":"data:2193",
                }
            }
```

@ESapenaVentura please review the `text` I used for the `content_description`.

File: broker/submissions/export_to_spreadsheet_service.py (hunk 1 above)
Comment (amnonkhen) [inline broker/submissions/export_to_spreadsheet_service.py]:
```suggestion
        self.ingest_api.post(f'/submission/{submission_uuid}/files/', spreadsheet_payload)
```

File: broker/submissions/export_to_spreadsheet_service.py (hunk 2 above)
Comment (ESapenaVentura) [reply broker/submissions/export_to_spreadsheet_service.py]:
```suggestion
            "file_core": {
                "file_name": spreadsheet_details.filename,
                "format": "xlsx",
                "file_source": "DCP/2 Ingest",
                "content_description": {
                    "text":"metadata spreadsheet",
                    "ontology":"data:2193",
                    "ontology_label":"Database entry metadata"
                }
            }
```

Comment (ESapenaVentura) [reply broker/submissions/export_to_spreadsheet_service.py]: @amnonkhen I would suggest this, it was missing the ontology_label. Other than that LGTM

**Comment by amnonkhen** [comment]
We also need to copy the spreadsheet to s3 so it is available to the exporter with the rest of the data files.
See [adding aws access to broker like the exporter](https://github.com/ebi-ait/ingest-kube-deployment/blob/1d71bc09ec4b398a65afaec218d3771ab88c0a8d/apps/ingest-exporter/templates/deployment.yaml#L56)

File: broker/submissions/export_to_spreadsheet_service.py [hunk 3]
```diff
@@ -33,23 +33,52 @@ def export_and_save(self, submission_uuid: str, storage_dir: str):
...
         finished_date = datetime.now(timezone.utc)
         self._patch(submission_url, create_date, finished_date)
 
-        self.logger.info(f'Done exporting spreadsheet for submission {submission_uuid}!')
+    def link_spreadsheet(self, spreadsheet_details, submission_uuid):
+        spreadsheet_payload = self.build_supplementary_file_payload(spreadsheet_details)
+        self.ingest_api.post(f'/submission/{submission_uuid}/files/', spreadsheet_payload)
+        submission = self.api.get_submission_by_uuid(submission_uuid)
+
+        self.ingest_api.post(f'/projects/')
```
Comment (amnonkhen) [inline broker/submissions/export_to_spreadsheet_service.py]:
```suggestion
        # TODO: find project id, build payload
        self.ingest_api.post(f'/projects/')
```

File: broker/submissions/export_to_spreadsheet_service.py [hunk 4]
```diff
@@ -33,23 +33,53 @@ def export_and_save(self, submission_uuid: str, storage_dir: str):
...
+
+    def update_spreadsheet_start(self, submission_url):
+        create_date = datetime.now(timezone.utc)
+        self._patch(submission_url, create_date)
+        return create_date
+
+    def build_supplementary_file_payload(self, spreadsheet_details):
+        return {
+            "describedBy": "https://schema.humancellatlas.org/type/file/2.5.0/supplementary_file",
```
Comment (MightyAx) [inline broker/submissions/export_to_spreadsheet_service.py:94]: the ingest client has a method for getting the latest schema of a certain type

File: broker/submissions/export_to_spreadsheet_service.py [hunk 5]
```diff
@@ -33,23 +33,53 @@ def export_and_save(self, submission_uuid: str, storage_dir: str):
...
         spreadsheet_details = self.get_spreadsheet_details(create_date, storage_dir, submission_uuid)
-
         workbook = self.export(submission_uuid)
+        self.save_spreadsheet(spreadsheet_details, workbook)
+        self.link_spreadsheet(spreadsheet_details, submission_uuid)
+        self.update_spreadsheet_finish(create_date, submission_url)
+        self.copy_to_s3_staging_area(spreadsheet_details, submission_uuid)
+        self.logger.info(f'Done exporting spreadsheet for submission {submission_uuid}!')
 
+    def save_spreadsheet(self, spreadsheet_details, workbook):
```
Comment (MightyAx) [inline broker/submissions/export_to_spreadsheet_service.py:68]: can be made static

**Review by MightyAx — CHANGES_REQUESTED** [review]
Also, lets update the broker's use of ingest-client to the latest version from pip, and use the pip-compile & pip-sync functions for dependancy resolution.

File: broker/submissions/export_to_spreadsheet_service.py [hunk 6]
```diff
@@ -33,23 +33,53 @@ def export_and_save(self, submission_uuid: str, storage_dir: str):
...
+        spreadsheet_payload = self.build_supplementary_file_payload(spreadsheet_details)
+        self.ingest_api.post(f'/submission/{submission_uuid}/files/', spreadsheet_payload)
+        submission = self.api.get_submission_by_uuid(submission_uuid)
+
+        # TODO: find project id, build payload
+        self.ingest_api.post(f'/projects/')
+
+    def update_spreadsheet_start(self, submission_url):
+        create_date = datetime.now(timezone.utc)
+        self._patch(submission_url, create_date)
```
Comment (MightyAx) [inline broker/submissions/export_to_spreadsheet_service.py:89]: Let's move this function into ingest client.

**Comment by MightyAx** [comment]
@amnonkhen  Resetting the base branches reveals that you are deleting the linter that was added in the previous PR but I don't know why

File: broker/submissions/export_to_spreadsheet_service.py (hunk 5 above)
Comment (amnonkhen) [reply broker/submissions/export_to_spreadsheet_service.py:68]: This entire class needs to be reorganised.

File: broker/submissions/export_to_spreadsheet_service.py (hunk 4 above)
Comment (amnonkhen) [reply broker/submissions/export_to_spreadsheet_service.py:94]: That is be awesome!

File: broker/submissions/export_to_spreadsheet_service.py (hunk 5 above)
Comment (amnonkhen) [reply broker/submissions/export_to_spreadsheet_service.py:68]: This is not urgent. This is a cleanup task. Let’s get the functionality working end to end.

File: broker/submissions/export_to_spreadsheet_service.py (hunk 6 above)
Comment (amnonkhen) [reply broker/submissions/export_to_spreadsheet_service.py:89]: This is not urgent. This is a cleanup task. Let’s get the functionality working end to end.

**Comment by MightyAx** [comment]
okay I'm closing the PR because of Git shenanigans.

Link to the new branch/PR here #44

**Pull request closed without merging**